src/foam/nanos/auth/AgentUserAuthService.java

```diff
@@ -0,0 +1,125 @@
+/**
+ * @license
+ * Copyright 2018 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.nanos.auth;
+
+import foam.core.ContextAwareSupport;
+import foam.core.X;
+import foam.dao.DAO;
+import foam.nanos.NanoService;
+import foam.nanos.session.Session;
+
+import static foam.mlang.MLang.EQ;
+import static foam.mlang.MLang.AND;
+
+public class AgentUserAuthService
+  extends    ContextAwareSupport
+  implements AgentAuthService, NanoService
+{
+  protected DAO userDAO_;
+  protected DAO groupDAO_;
+  protected DAO sessionDAO_;
+  protected DAO agentJunctionDAO_;
+
+  public AgentUserAuthService(X x) {
+    setX(x);
+  }
+
+  @Override
+  public void start() {
+    userDAO_     = (DAO) getX().get("localUserDAO");
+    groupDAO_    = (DAO) getX().get("groupDAO");
+    sessionDAO_  = (DAO) getX().get("sessionDAO");
+    agentJunctionDAO_  = (DAO) getX().get("agentJunctionDAO");
+  }
+
+  /* 
+    Sets the currently logged in user as an "agent" within the context &
+    sets "user" in the context to the passed in user. This allows users to
+    act on behalf of others while retaining information on the user.
+  */
+  public User actAs(X x, User superUser) throws AuthenticationException {
+    User agent = (User) x.get("user");
+    User user = (User) userDAO_.find(superUser.getId());
```

Comment:
Since this method accepts a context as a parameter, I think we should be calling context-oriented methods where we can. Please use the context-oriented `find_` here and fix all places below as well.

Reply:
With regard to using the passed context, subsequent calls should be of the form:
userDAO_.inX(x).find(superUser.getId());